File: javascript/src/breakdown.ts
import Decimal from "decimal.js";
import { quantizeMoney } from "./expr.js";

/**
 * Granular credit cost report for a usage event.
 *
 * All money fields are `Decimal` (never binary `number`) and are quantized to
 * 4dp ROUND_HALF_UP. `total` is the single source of truth: it is recomputed
 * from the components (clamped at 0) and quantized — there is NO implicit
 * integer truncation of the total (that was the revenue-leak bug).
 */
export interface CostBreakdown {
  modelCredits: Decimal;
  toolCredits: Decimal;
  searchCredits: Decimal;
  cacheSavings: Decimal;
  fixedCredits: Decimal;
  total: Decimal;
  breakdown: Record<string, unknown>;
}

export function makeCostBreakdown(partial?: {
  modelCredits?: Decimal;
  toolCredits?: Decimal;
  searchCredits?: Decimal;
  cacheSavings?: Decimal;
  fixedCredits?: Decimal;
  breakdown?: Record<string, unknown>;
}): CostBreakdown {
  const modelCredits = quantizeMoney(partial?.modelCredits ?? new Decimal(0));
  const toolCredits = quantizeMoney(partial?.toolCredits ?? new Decimal(0));
  const searchCredits = quantizeMoney(partial?.searchCredits ?? new Decimal(0));
  const cacheSavings = quantizeMoney(partial?.cacheSavings ?? new Decimal(0));
  const fixedCredits = quantizeMoney(partial?.fixedCredits ?? new Decimal(0));

  // Single source of truth for the total: sum of components, clamped at 0,
  // quantized to 4dp HALF_UP.
  const rawTotal = modelCredits
    .plus(toolCredits)
    .plus(searchCredits)
    .plus(fixedCredits)
    .plus(cacheSavings);
  const total = quantizeMoney(Decimal.max(new Decimal(0), rawTotal));

  return {
    modelCredits,
    toolCredits,
    searchCredits,
    cacheSavings,
    fixedCredits,
    total,
    breakdown: partial?.breakdown ?? {},
  };
}

File: python/src/ducto/breakdown.py
"""Aggregated cost breakdown produced by ``PricingEngine.calculate()``.

The ``CostBreakdown`` model holds per-category credit costs (all
:class:`decimal.Decimal`, quantized to 4 dp ROUND_HALF_UP) and a ``total``.

Single source of truth (M3): ``total`` is computed **once**, by the engine,
and passed in. The model no longer recomputes/overwrites it in a validator,
so there is exactly one place that decides clamping + rounding.
"""

from __future__ import annotations

from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class CostBreakdown(BaseModel):
    """Granular credit cost report for a usage event or batch.

    All monetary fields are :class:`decimal.Decimal`. The engine quantizes
    every field to 4 dp ROUND_HALF_UP and clamps ``total`` to ``>= 0`` before
    constructing this model; nothing here re-derives those numbers.
    """

    model_config = ConfigDict(extra="forbid")

    model_credits: Decimal = Decimal("0.0000")
    tool_credits: Decimal = Decimal("0.0000")
    search_credits: Decimal = Decimal("0.0000")
    cache_savings: Decimal = Decimal("0.0000")
    fixed_credits: Decimal = Decimal("0.0000")
    total: Decimal = Decimal("0.0000")
    breakdown: dict[str, Any] = Field(default_factory=dict)
